src/sentry/workflow_engine/models/detector.py

```diff
@@ -197,6 +197,54 @@ def get_state_data(self, group_keys: list[str | None]) -> dict[str | None, Detec
             )
         return results
 
+    def evaluate(self, data_packet: DataPacket[T]) -> list[DetectorEvaluationResult]:
+        """
+        Evaluates a given data packet and returns a list of `DetectorEvaluationResult`.
+        There will be one result for each group key result in the packet, unless the
+        evaluation is skipped due to various rules.
+        """
+        dedupe_value = self.get_dedupe_value(data_packet)
+        group_values = self.get_group_key_values(data_packet)
+        all_state_data = self.get_state_data(group_values.keys())
+        results = []
+        for group_key, group_value in group_values.items():
+            result = self.evaluate_group_key_value(
+                group_key, group_value, all_state_data[group_key], dedupe_value
+            )
+            if result:
+                results.append(result)
+        return results
+
+    def evaluate_group_key_value(
+        self, group_key: str, value: int, state_data: DetectorStateData, dedupe_value: int
+    ) -> DetectorEvaluationResult | None:
+        """
+        Evaluates a value associated with a given `group_key` and returns a `DetectorEvaluationResult` with the results
+        and any state changes that need to be made.
+
+        Checks that we haven't already processed this datapacket for this group_key, and skips evaluation if we have.
+        """
+        if dedupe_value <= state_data.dedupe_value:
+            # TODO: Does it actually make more sense to just do this at the data packet level rather than the group
+            # key level?
```

Comment:
🤔 not sure where this would happen on a data packet -- that said i do like this as a metric in general.

although, this raises an interesting question, is there any de-duping that we should add to process data sources? (every code path i've mapped end to end only provide a single data_packet so it's a little harder for me to tell)